sample/src/main/kotlin/io/getstream/chat/sample/feature/login/LoginFragment.kt

```diff
@@ -6,63 +6,87 @@ import androidx.core.view.isVisible
 import androidx.fragment.app.Fragment
 import androidx.lifecycle.Observer
 import androidx.navigation.fragment.findNavController
+import com.getstream.sdk.chat.BuildConfig
 import io.getstream.chat.sample.R
 import io.getstream.chat.sample.application.EXTRA_CHANNEL_ID
 import io.getstream.chat.sample.common.navigateSafely
 import io.getstream.chat.sample.common.showToast
-import io.getstream.chat.sample.data.user.User
 import kotlinx.android.synthetic.main.fragment_login.*
+import kotlinx.android.synthetic.main.fragment_login.loadingProgressBar
 import org.koin.androidx.viewmodel.ext.android.viewModel
 
 class LoginFragment : Fragment(R.layout.fragment_login) {
 
     private val viewModel: LoginViewModel by viewModel()
 
-    private val adapter = UsersListAdapter {
-        viewModel.userClicked(this)
-    }
-
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
-        usersList.adapter = adapter
+        sdkVersion.text = getString(R.string.login_sdk_version_template, BuildConfig.VERSION_NAME)
+        developmentUsers.setOnClickListener {
+            redirectToUsersScreen()
+        }
+        loginButton.setOnClickListener {
+            val credentials = collectCredentials()
+            if (validateCredentials(credentials)) {
+                viewModel.loginButtonClicked(credentials)
+            } else {
+                showToast(getString(R.string.login_validation_error))
```

Comment:
Agree, unfortunately the design in Figma is not final and it is being reworked at the moment. But yes a persistent errors would be better. Don't want to add SingleLiveData to the viewmodel just to show validation events with toasts. We need to be careful with mixing "state" (loading, content, error, etc.) and one shot "events" (displaying toasts, etc.) in a livedata.

Reply:
I think `TextInputLayout.setError()` is widely used and standard way of showing input / validation error on Android and could be used instead of toast.

IMO, validation errors like other one-shot events should be emitted by view model as well. It's a piece of logic to validate input data and decide if we can move forward with some backend call (emit `State.Loading`) or abort and display validation error to the user (emit `State.InvalidData`).